Refactoring of alphabets

h-2 opened the pull request:
## major refactorisation of all alphabets

- there are now alphabet base classes that our alphabets inherit from, this reduces code duplication
- alphabets are no longer aggregate types fixing some weird issues
- our alphabets are still PODs although this is not required by the alphabet_concept anymore
- `constexpr` default construction has been fixed everywhere
- cartesian composition no longer inherits pod_tuple, but actually compresses into a single rank value
  - `gapped<dna5>` now consumes one byte, not two
  - all operations are now constant time, not linear in the number of elements
- introduction of alphabet_proxy, a generic proxy base that also models the concepts of it's emulated type
- fixed a heap of bugs I found on the way

## Reviews:

- @joshuak94 can you review the changes to the mask alphabet and gap?
- @sarahet can you review the changes regarding the amino acids? It's not much ;)
- @joergi-w can you review the changes to the structure alphabets and the dna alphabets?
- @mariehoffmann can you review the changes to the qualities?
- @smehringer @marehr Can you review rest, in particular the changes to cartesian_composition and the alphabet_proxy?

Follow-Up TODOs:
  - [ ] investigate the two ICEs
  - [ ] make union_composition require the same char_type on alternatives
  - [ ] rename `composition` module to `composite`
  - [ ] rename `cartesian_composition` to `alphabet_tuple`
  - [ ] rename `union_composition` to `alphabet_variant`
  - [ ] implement `alphabet_any`
  - [ ] switch all alphabets from enum-notation to using a char literal (only done for nucleotides)

h-2 commented:
Wow, I have the same GCC7 verison as travis, but the compiler on travis has serious issues :thinking:

h-2 commented:
> Wow, I have the same GCC7 verison as travis, but the compiler on travis has serious issues

All tests pass locally, I will investigate the wuss problem tomorrow. @joergi-w I remember we had this before... how did we solve it?

Anyway, this is good to be reviewed.

h-2 commented:
@marehr Do you know if the travis GCC still has some non-default stuff?

marehr commented:
Except that the build might be on some weird gcc state and that the std string implementation is the wrong one. No.

joergi-w commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -53,10 +51,14 @@
  * \implements seqan3::nucleotide_concept
  * \implements seqan3::detail::constexpr_alphabet_concept
+ * \implements seqan3::trivial_concept
+ * \implements seqan3::standar_layout_concept
```
`standard_layout_concept`
This typo occurs very often, please do a search.

smehringer commented on include/seqan3/alphabet/aminoacid/aa20.hpp:
```diff
@@ -81,180 +81,32 @@ namespace seqan3
  */
-
-struct aa20
+class aa20 : public aminoacid_base<aa20, 27>
```
why is the second template parameter (size) 27 for aa20?

h-2 replied:
good catch

smehringer commented on include/seqan3/alphabet/detail/alphabet_base.hpp:
```diff
@@ -0,0 +1,237 @@
+ *   * `static std::array<rank_type, 256> constexpr char_to_rank` that defines for every possible character value the
+ *     corresponding rank value.
+ *
+ * TODO add snippet
```
TODO

smehringer commented on include/seqan3/alphabet/detail/alphabet_base.hpp:
```diff
@@ -0,0 +1,237 @@
+    constexpr alphabet_base(alphabet_base &&) = default;
+    constexpr alphabet_base & operator=(alphabet_base const &) = default;
+    constexpr alphabet_base & operator=(alphabet_base &&) = default;
+    //!\}
```
The destructor is missing

h-2 replied:
I can add it everywhere, but I am not sure if we really need this...

smehringer commented on include/seqan3/alphabet/detail/alphabet_base.hpp:
```diff
@@ -0,0 +1,237 @@
+        return derived_type::rank_to_char[rank];
+    }
+
+    /*!\brief Return the letter's numeric value or rank in the alphabet.
```
`Return the letter's numeric value (rank) in the alphabet.` :nail_care: 
I understood it returns either or but it was meant as a synonym correct?

smehringer commented on include/seqan3/alphabet/detail/alphabet_base.hpp:
```diff
@@ -0,0 +1,237 @@
+     */
+    constexpr derived_type & assign_char(std::conditional_t<std::Same<char_type, void>, char, char_type> const c) noexcept
+    //!\cond
+        requires !std::Same<char_type, void>
```
If you explicitly require that the type is not void, why do you need the `std::conditional` (or the other way around)

h-2 replied:
Because if the parameter type is void, the program becomes ill-formed (even though the function would never be chosen due to the concept check). I know, it's not pretty.

smehringer commented on include/seqan3/alphabet/detail/alphabet_base.hpp:
```diff
@@ -0,0 +1,237 @@
+    constexpr derived_type & assign_rank(rank_type const c) noexcept
+    {
+        assert(c < value_size);
+        rank = c;
```
what happens in release mode here if the assertion does not catch?

h-2 replied:
Undefined Behaviour.

smehringer commented on include/seqan3/alphabet/nucleotide/nucleotide_base.hpp:
```diff
@@ -0,0 +1,160 @@
+     */
+//     //!\brief Implicit conversion between dna* and rna* of the same size.
+//     //!\tparam other_nucl_type The type to convert to; must satisfy seqan3::nucleotide_concept and have the same \link value_size \endlink.
+//     template <typename other_nucl_type>
```
If this is part of a follow up issue reference the issue in a comment

h-2 replied:
No, this can be removed!

smehringer commented on include/seqan3/alphabet/nucleotide/nucleotide_base.hpp:
```diff
@@ -0,0 +1,160 @@
+ * \tparam size         The size of the alphabet.
+ * \tparam char_t       The character type of the alphabet (set this to `void` when defining just a
+ *                      seqan3::semi_alphabet_concept).
+ */
```
```
 *
 * You can use this class to define your own nucleotide alphabet, but types are not required to be based on it to model
 * seqan3::nucleotide_concept, it is purely a way to avoid code duplication.
 *
 * The derived type needs to define only the following table as static member variables:
 *
 *   * `static ........ complement_table` that defines for every possible rank value
 *     the corresponding complement.
```

smehringer commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+ */
+
+#pragma once
+
```
remove double newline :nail_care:

smehringer commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+    /*!\name Read functions
+     * \{
+     */
+    constexpr char_type to_char() const noexcept
```
add `\copydoc` from base class

smehringer commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+        return static_cast<char_type>(to_rank()) + derived_type::offset_char;
+    }
+
+    constexpr phred_type to_phred() const noexcept
```
documentation missing

smehringer commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+
+    //!\brief Explicit conversion to any other quality alphabet (via char representation).
+    //!\tparam other_qual_type The type to convert to; must satisfy seqan3::quality_concept.
+    //
```
remove

smehringer commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+        {
+            std::array<rank_type, 256> ret{};
+
+            phred_type min = derived_type::offset_phred;
```
Ist das const oder kann das weg? :grin:

smehringer commented on include/seqan3/io/stream/char_operations.hpp:
```diff
@@ -0,0 +1,129 @@
+    {
+        std::array<char_type, s> ret{};
+
+        for (size_t i = 0; i < s; ++i)
```
Why are you using `size_t` if you need to cast to `char_type` when assigning `i`? Why don't you use `char_type` directly?

smehringer commented on include/seqan3/io/stream/char_operations.hpp:
```diff
@@ -0,0 +1,129 @@
+
+//!\brief Auxilliary table for seqan3::to_lower.
+template <typename char_type, size_t s = static_cast<size_t>(std::pow(2, sizeof(char_type) * 8))>
+inline std::array<char_type, s> constexpr to_lower_table
```
drop a comment that `s` is only a template parameter for convenience but should not be specified manually

smehringer commented on include/seqan3/core/concept/core_language.hpp:
```diff
@@ -183,4 +196,17 @@ template <typename t>
+ *
+ * \details
+ *
+ * Note that this requires less than std::Assignable!
```
Add `It simply tests if the expression std::declval<T>() = std::declval<U>() is well-formed.`

smehringer commented on include/seqan3/core/concept/core_language.hpp:
```diff
@@ -143,6 +143,19 @@ template <typename t>
+//!\cond
+
+template <typename t>
+concept char_concept = std::Integral<t> &&
```
isn't requiring `std::Integral<t>` obsolete if you test for the exact type afterward?

h-2 replied:
But I want char_concept to more refined than Integral so you can use these for specialisation.

smehringer commented on include/seqan3/io/alignment_file/detail.hpp:
```diff
@@ -251,8 +252,11 @@ std::string get_cigar_string(alignment_type && alignment,
 {
-    return get_cigar_string(std::get<0>(alignment), std::get<1>(alignment),
-                            query_start_pos, query_end_pos, extended_cigar);
+    return get_cigar_string(get<0>(alignment),
```
maybe remove this NOP change

smehringer commented on test/unit/alignment/aligned_sequence_test.cpp:
```diff
@@ -154,7 +154,10 @@ TYPED_TEST(aligned_sequence_test, cigar_string)
         initialize_typed_test_container(ref,  "ACGTGAT--CTG");
         initialize_typed_test_container(read, "ACGT-CGTAGTG");
 
+//         [[maybe_unused]] char c = detail::compare_aligned_values(ref[0], read[0], false);
```
why is this commented out?

h-2 replied:
debug stuff that I forgot to remove.

smehringer commented on test/unit/alphabet/alphabet_test.cpp:
```diff
@@ -66,17 +85,18 @@ using alphabet_types = ::testing::Types<dna4, dna5, dna15, rna4, rna5, rna15,
                                         gapped<dna15>,
                                         char, char16_t, // char32_t, too slow
                                         uint8_t, uint16_t, // uint32_t, too slow
-                                        dna4q,
```
maybe remove this unnecessary change again. Or did this actually make a difference? ^^

h-2 replied:
well dna4q is just another name for `qualified<dna4>` so it makes sense to group it together. But I assume this will change anyway if #535 is merged before this one.

marehr replied:
I thought this one is #535 :thinking:

smehringer commented on include/seqan3/alphabet/nucleotide/all.hpp:
```diff
@@ -88,12 +88,14 @@
  *
  * The nucleotide submodule defines seqan3::nucleotide_concept which encompasses all the alphabets defined in the
- * submodule and refines seqan3::alphabet_concept.
+ * submodule and refines seqan3::alphabet_concept. The only additional requirement of types is that they can be
```
`... is that their values can be complemented`

smehringer commented on include/seqan3/alphabet/nucleotide/all.hpp:
```diff
@@ -105,21 +107,11 @@
- * |   ...          |    ...         |     ---      |    ---       |    ...         |     ---      |     ---      |
- * | type::N        |    `'N'`       |    `'N'`     |    ---       |    `'N'`       |    `'N'`     |     ---      |
- * | type::UNKNOWN  |    `'N'`       |    `'N'`     |    `'A'`     |    `'N'`       |    `'N'`     |    `'A'`     |
+ * The character values can be printed by `to_char()`.
```
`the to_char member function`

h-2 replied:
No, also and especially the free function`to_char`!

smehringer replied:
then `seqan3::to_char` to be explicit?

smehringer commented on include/seqan3/alphabet/nucleotide/all.hpp:
```diff
@@ -105,21 +107,11 @@
  *
- * All alphabets define at least the letters `::%A`, `::%C`, `::%G`, `::%T`, `::%U`, `::%UNKNOWN` (although some
- * might be aliases of others).
+ * All alphabets define all letters, although some resolve to the same internal values.
```
`.... E.g. for seqan3::dna4 all values other than A,C,G,T resolve to A.`

smehringer commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -54,10 +51,14 @@
 namespace seqan3
 {
 
+struct rna5;
```
If `dna4` is a class now why is `rna4` a struct still?

h-2 replied:
This is just a forward and it doesn't matter for forwards.

smehringer commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -69,174 +70,72 @@ namespace seqan3
             // initialize with UNKNOWN (std::array::fill unfortunately not constexpr)
             for (auto & c : ret)
-                c = in_t::UNKNOWN;
+                c = 3; // == 'N'
```
I would use `base_t::UNKOWN`just in case we do change the order again..

h-2 replied:
The problem is that this is not yet fully defined. I would have to initialise the struct outside to do this. But it's not too big a problem I think, because this function is now the only place where the actual relation of char and rank number is defined so if we change it, we definetely touch this function anyway!

smehringer commented on include/seqan3/alphabet/nucleotide/nucleotide_base.hpp:
```diff
@@ -0,0 +1,160 @@
+ *                      seqan3::semi_alphabet_concept).
+ */
+template <typename derived_type, auto size, typename char_t = char>
+class nucleotide_base : public alphabet_base<derived_type, size, char_t>
```
As discussed, remove template parameter `char_t` otherwise we may run into problems when someone specifies a bigger char type but our `char_to_rank` table is an array of size 265.

smehringer commented on include/seqan3/alphabet/nucleotide/rna15.hpp:
```diff
@@ -56,112 +52,58 @@ namespace seqan3
+    constexpr rna15(rna15 const &) = default;
+    constexpr rna15(rna15 &&) = default;
+    constexpr rna15 & operator=(rna15 const &) = default;
+    constexpr rna15 & operator=(rna15 &&) = default;
```
The destructor is missing

smehringer commented on include/seqan3/alphabet/nucleotide/dna5.hpp:
```diff
@@ -69,174 +70,72 @@ namespace seqan3
+    constexpr dna5(t const & r)
+    {
+        using seqan3::to_rank;
+        assign_rank(to_rank(r));
```
for`dna15` you use the member function to_rank instead which seems sufficient to me since we know the type. I would change this to be consistent :nail_care:

smehringer commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -248,69 +125,35 @@ struct dna15
-            ret['H'] = in_t::H; ret['h'] = in_t::H;
-            ret['V'] = in_t::V; ret['v'] = in_t::V;
-            ret['N'] = in_t::N; ret['n'] = in_t::N;
+                c = 8; // rank of 'N'
```
again use `base_t::UNKOWN` if possible

smehringer commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -69,218 +70,44 @@ namespace seqan3
+    constexpr dna4(dna4 const &) = default;
+    constexpr dna4(dna4 &&) = default;
+    constexpr dna4 & operator=(dna4 const &) = default;
+    constexpr dna4 & operator=(dna4 &&) = default;
```
destructor missing

smehringer commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -56,113 +53,76 @@ namespace seqan3
+    constexpr rna4(rna4 const &) = default;
+    constexpr rna4(rna4 &&) = default;
+    constexpr rna4 & operator=(rna4 const &) = default;
+    constexpr rna4 & operator=(rna4 &&) = default;
```
destructor is missing

smehringer commented on include/seqan3/alphabet/nucleotide/rna4.hpp:
```diff
@@ -56,113 +53,76 @@ namespace seqan3
-    }
-    //!\}
+protected:
+    //!\privatesection
```
do you need this?

h-2 replied:
Well I don't want things to appear in the user doc.

smehringer commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -56,116 +53,78 @@ namespace seqan3
+    constexpr rna5(rna5 const &) = default;
+    constexpr rna5(rna5 &&) = default;
+    constexpr rna5 & operator=(rna5 const &) = default;
+    constexpr rna5 & operator=(rna5 &&) = default;
```
destructor missing

smehringer commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -56,116 +53,78 @@ namespace seqan3
-constexpr rna5 rna5::UNKNOWN{rna5::N};
+constexpr std::array<rna5, rna5::value_size> rna5::complement_table
+{
+    rna5::T,    // complement of rna5::A
```
shouldn't this be U for consistency/readability? although internally they are the same? If not can't we simply copy the dna5 complement table?

h-2 replied:
I'll have a look!

smehringer commented on include/seqan3/alphabet/gap/gap.hpp:
```diff
@@ -117,32 +117,32 @@ struct gap
     //!\name Comparison operators
     //!\{
-    constexpr bool operator==(gap const &) const noexcept
+    friend constexpr bool operator==(gap const &, gap const &) noexcept
```
remove reference

h-2 replied:
This doesn't make a difference, because there is no parameter anyway. I assume this function will always be replaced with just the return value in the code.

h-2 replied:
(We haven't discussed this for SeqAn3, because luckily we don't have so many instances of "dead" parameters, but I think `foo const &` in this case makes it even easier to see that there is no parameter name than just `foo`)

smehringer commented on include/seqan3/alphabet/aminoacid/aminoacid_base.hpp:
```diff
@@ -0,0 +1,139 @@
+ * \tparam char_t       The character type of the alphabet (set this to `void` when defining just a
+ *                      seqan3::semi_alphabet_concept).
+ */
+template <typename derived_type, auto size, typename char_t = char>
```
remove the `char_t` template parameter as you assume `char` in `char_to_rank` table

h-2 commented:
:tada: I have worked around the travis issues with wuss. Now only building `view_translation_test` fails because it needs 8GB to build for some reason :thinking:

rrahn commented:
@h-2 please give me ping when all comments have been worked on. I can then have a final review.

smehringer commented on test/unit/alphabet/composition/cartesian_composition_test.cpp:
```diff
@@ -307,23 +313,11 @@ TYPED_TEST(cartesian_composition_test, get_i)
-
-    static_assert(std::is_same_v<decltype(std::get<0>(t0)), decltype(TestFixture::value_1()) &>);
-    static_assert(std::is_same_v<decltype(std::get<1>(t0)), decltype(TestFixture::value_2()) &>);
+//     static_assert(std::is_same_v<decltype(get<0>(t0)), decltype(TestFixture::value_1()) &>);
```
why is this commented out?

h-2 commented on include/seqan3/alphabet/aminoacid/aminoacid_base.hpp:
```diff
@@ -0,0 +1,139 @@
+    //!\}
+
+    //!\brief Befriend the derived class so it can instantiate.
+    friend derived_type;
```
hm, in general yes, but in this case befriending the derived type only happens, because the constructors are private so I thought it is better to put it after the constructors...

h-2 commented on include/seqan3/alphabet/aminoacid/aminoacid_base.hpp:
```diff
@@ -0,0 +1,139 @@
+     * \details Similar to an Enum interface.
+     */
+    //!\{
+    static derived_type constexpr A          = assign_char(derived_type{}, 'A');
```
It was always const and always constexpr. Before I had to use the trick of declaring it const, but defining it constexpr outside the class. Now I declare it of the derived_type so I can define it constexpr immediately (and initialise inside the class).

smehringer commented on include/seqan3/alphabet/composition/union_composition.hpp:
```diff
@@ -84,11 +87,23 @@ namespace seqan3
-    requires (detail::constexpr_alphabet_concept<alternative_types> && ... && (sizeof...(alternative_types) >= 2))
+    requires (detail::constexpr_alphabet_concept<alternative_types> && ...) &&
+             (sizeof...(alternative_types) >= 2)
+             //TODO same char_type
```
TODO

smehringer replied:
can you reference it in the comment then? I think this will always make synchronization easier

h-2 commented on include/seqan3/core/concept/tuple.hpp:
```diff
@@ -76,11 +76,14 @@ concept tuple_get_concept = requires (tuple_t & v, tuple_t const & v_c)
-    { std::get<0>(v_c)            } -> typename std::tuple_element<0, tuple_t>::type const &;
-    { std::get<0>(std::move(v))   } -> typename std::tuple_element<0, tuple_t>::type &&;
+    { get<0>(v)              } -> typename std::tuple_element<0, tuple_t>::type;
+//     requires weakly_assignable_concept<decltype(get<0>(v)), typename std::tuple_element<0, tuple_t>::type>;
```
Yeah, this whole concept is really broken. I need to sit down with @rrahn and re-evaluate how we want to do this in the future.

smehringer commented on include/seqan3/alphabet/composition/union_composition.hpp:
```diff
@@ -202,9 +214,11 @@ class union_composition
-        _value{rank_by_type_(alternative)}
-    {}
+    constexpr union_composition(alternative_t const & alternative) noexcept
+
```
remove newline

h-2 commented on include/seqan3/alphabet/nucleotide/concept.hpp:
```diff
@@ -62,11 +63,9 @@ namespace seqan3
-    requires alphabet_concept<type>;
-
-    { complement(v) } -> std::remove_reference_t<type>;
+    requires std::Same<decltype(complement(v)),        decltype(c)>;
```
`->` means convertible to. I am checking for identity now.

marehr replied:
Why not `-> std::Same<std::remove_reference_t<type>>`?

h-2 commented on include/seqan3/alphabet/detail/alphabet_proxy.hpp:
```diff
@@ -0,0 +1,291 @@
+    //!\brief Construction from the emulated type.
+    constexpr alphabet_proxy(alphabet_type const a)
+    {
+        base_t::assign_rank(seqan3::to_rank(a));
```
Because I know the base_t has a member function and I always want to call that. There is no need to lookup functions in other namespaces.

smehringer replied:
why don't you use the member function directly then?

h-2 replied:
I am  :laughing:

h-2 commented on include/seqan3/alphabet/detail/alphabet_proxy.hpp:
```diff
@@ -0,0 +1,291 @@
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr alphabet_proxy() : base_t{} {}
```
https://isocpp.org/blog/2017/04/quick-q-prevent-user-from-derive-from-incorrect-crtp-base

(I think we discussed this a few times already :wink: )

smehringer replied:
yes :see_no_evil: I keep forgetting..

smehringer commented on test/unit/alphabet/mask/masked_test.cpp:
```diff
@@ -198,38 +198,18 @@
     masked<dna4> t0{dna4::C, mask::MASKED};
 
-    static_assert(std::is_same_v<decltype(seqan3::get<0>(t0)), dna4 &>);
+//     static_assert(std::is_same_v<decltype(get<0>(t0)), dna4 &>);
```
why is this commented out ?

h-2 commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -289,57 +116,42 @@ struct dna4
+            // reverse mapping for characters and their lowercase
+            for (rank_type rnk = 0u; rnk < value_size; ++rnk)
+            {
+                ret[static_cast<rank_type>(         rank_to_char[rnk]) ] = rnk;
```
I'll have a look at this again!

h-2 commented on include/seqan3/alphabet/quality/phred42.hpp:
```diff
@@ -67,217 +66,40 @@ namespace seqan3
  */
-struct phred42
+class phred42 : public quality_base<phred42, 42>
 {
```
They are classes, because they are not aggregate types anymore and some have extra constructors. This fixes some long-standing design problems and gives us the opportunity to add custom constructors. The memory behaviour is still the same, because they are still POD, as you said.

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+
+            // map too small to smallest possible
+            for (int64_t i = std::numeric_limits<char_type>::lowest(); i < min ; ++i)
+                ret[static_cast<rank_type>(i)] = - derived_type::offset_char + min;
```
I wanted to make it consistent with above, but I agree it is not very readable, I will have another look at improving this!

h-2 commented:
> This is really weird, because the second test tests everything alphabet related and it only increases a little, while the first one has an almost unchanged codebase and increases dramatically :see_no_evil:

I have added a fix that not only reverts the increase in compile-time and memory usage, but actually lowers it significantly: from 3.5GiB(8GiB) to 1.3GiB and time from 37s(51) to 20s.

smehringer commented:
Question: we default the underlying char type of out alphabets to be `char` but maybe we should default it to `unsigned char` to avoid the platform dependent `signed/unsigned` issue?

h-2 commented:
@smehringer I thought about this for a while, as well, but I came to the conclusion that we should stick with `char`, because since the assign_char function will usually get a plain `char` as argument (IO or conversion from string) it will need to do an extra conversion.

h-2 commented on include/seqan3/alphabet/composition/cartesian_composition.hpp:
```diff
@@ -167,37 +183,109 @@ class cartesian_composition : public pod_tuple<first_component_type, component_t
-                                               alphabet_size_v<component_types>)};
+    //!\brief Export this type's components in a visible manner.
+    //!\private
+    using seqan3_tuple_element_types = component_list;
```
It is a tuple, just the internal representation is different. The thing with `seqan3_` is a hack because we change the standard library customisation point for tuples based on this and I don't want that picking up on some other program's possible member type "tuple_element_types".
It's not pretty, but I don't have a better solution right now, maybe @rrahn knows something in his final review.

smehringer commented on include/seqan3/alphabet/nucleotide/nucleotide_base.hpp:
```diff
@@ -47,17 +47,26 @@ namespace seqan3
+ *
+ * The derived type needs to define only the following table as static member variable:
+ *
+ *   * `static ........ complement_table` that defines for every possible rank value
```
:rofl: please exchange the `.......` with an actual type

h-2 commented:
This is now "done" from my POV. It would be great if you could review asap, this touches a lot, especially with the change to character literals and the removal of the literal namespace.

h-2 commented on include/seqan3/alphabet/nucleotide/rna5.hpp:
```diff
@@ -56,152 +53,108 @@ namespace seqan3
-        'U',
-        'N'
+        'N',
+        'U'
```
You can never rely on the rank, it is implementation defined. But we have changed this consistently now to always be in alphabetical order. This makes debugging a lot easier, because the sorting order is the natural order and also evaluating binary values is much easier.

h-2 commented on include/seqan3/alphabet/structure/dot_bracket3.hpp:
```diff
@@ -276,14 +192,13 @@ namespace seqan3::literal
  *
  * \attention
- * All seqan3 literals are in the namespace seqan3::literal!
+ * All seqan3 literals are in the namespace seqan3!
```
all non-nucleotide alphabet still need to be adjusted to have use character literals. We can adjust all of the related documentation as part of that change (but it won't happen in this PR).

h-2 commented:
@smehringer @joergi-w @mariehoffmann I think I have addressed all your comments except some minor things in the tests. Can you have a look and mark respective comments as resolved?

@marehr I can haz review plz? :cat:

rrahn commented on include/seqan3/alphabet/composition/cartesian_composition.hpp:
```diff
@@ -321,240 +444,185 @@ class cartesian_composition : public pod_tuple<first_component_type, component_t
+    friend constexpr auto get(cartesian_composition & l)
     {
-        return to_rank_impl(positions);
+        using t = meta::at_c<component_list, index>;
```
using an index out of range will trigger an infintiy compiler loop here. I think I encountered similar problems. But maybe it was another meta:: functionality. Just check this maybe.

rrahn replied:
So you should add a static_assert

marehr commented on include/seqan3/core/tuple_utility.hpp:
```diff
@@ -199,3 +199,26 @@ constexpr auto tuple_pop_front(tuple_t && t)
+//!\brief Value metafunction that returns whether a type is an element type of a tuple. [Auxiliary overload]
+template <typename element_t, tuple_like_concept tuple_t, size_t ...idx>
+inline bool constexpr is_element_of<element_t, tuple_t, std::index_sequence<idx...>> =
+    (std::Same<element_t, std::tuple_element_t<idx, tuple_t>> || ...);
```
to suppo
```c++
std::Same<element_t, std::tuple_element_t<idx, tuple_t>> || ...
```
to
```c++
std::Same<element_t, std::tuple_element_t<idx, tuple_t>> || ... || true
```

h-2 replied:
?

marehr replied:
My comment was truncated: to support empty tuples

marehr commented on include/seqan3/core/tuple_utility.hpp:
```diff
@@ -199,3 +199,26 @@ constexpr auto tuple_pop_front(tuple_t && t)
+inline bool constexpr is_element_of<element_t, tuple_t> =
+    is_element_of<element_t, tuple_t, std::make_index_sequence<std::tuple_size_v<tuple_t>>>;
+
+} // namespace seqan3::detai
```
seqan3::detai**l**

h-2 replied:
gone

marehr commented on include/seqan3/core/concept/tuple.hpp:
```diff
@@ -76,11 +76,14 @@ concept tuple_get_concept = requires (tuple_t & v, tuple_t const & v_c)
-    { std::get<0>(v)              } -> typename std::tuple_element<0, tuple_t>::type &;
-    { std::get<0>(v_c)            } -> typename std::tuple_element<0, tuple_t>::type const &;
-    { std::get<0>(std::move(v))   } -> typename std::tuple_element<0, tuple_t>::type &&;
+    { get<0>(v)              } -> typename std::tuple_element<0, tuple_t>::type;
```
Could we add `namespace stl; get<0>(v)` for adl support?

h-2 replied:
get without namespace already does ADL. You cannot include namespaces inside constraint declarations AFAIK.

marehr commented on include/seqan3/alphabet/detail/alphabet_proxy.hpp:
```diff
@@ -0,0 +1,292 @@
+{
+private:
+    //!\brief Type of the base class.
+    using base_t =  alphabet_base<derived_type,
```
`base_type`

marehr commented on include/seqan3/alphabet/detail/alphabet_proxy.hpp:
```diff
@@ -0,0 +1,292 @@
+    constexpr derived_type & operator=(alphabet_type const & c) noexcept
+    {
+        base_t::assign_rank(seqan3::to_rank(c));
+        static_cast<derived_type &>(*this).impl(); // <- this invokes the actual proxy behaviour!
```
`.impl` is a confusing name. What does that do?

marehr commented on include/seqan3/alphabet/detail/alphabet_proxy.hpp:
```diff
@@ -0,0 +1,292 @@
+    //!\brief Assignment from any type that the emulated type is assignable from.
+    template <typename indirect_assignable_type>
+    constexpr derived_type & operator=(indirect_assignable_type const & c) noexcept
+        requires weakly_assignable_concept<alphabet_type, indirect_assignable_type>
```
why is this requires after the function signature and does this not need to be guarded by `//!\cond`?

h-2 replied:
Most of the times doxygen copes with trailing requires statements well so they don't need to be escaped. Only more complex stuff needs escaping in both cases. We realised this early on and did this in most other places, too, e.g.:
https://github.com/seqan/seqan3/blob/master/include/seqan3/alphabet/detail/member_exposure.hpp

marehr replied:
Thank you, good to know!

marehr commented on include/seqan3/range/container/bitcompressed_vector.hpp:
```diff
@@ -125,12 +131,21 @@ class bitcompressed_vector
         ranges::semiregular_t<internal_proxy_type> internal_proxy;
 
+        //!\brief Update the sdsl-proxy.
+        constexpr void impl()
```
why not call it `update_proxy`? or `propagate_value`?

This seems to be an event, in different languages I would write `on_update` xD

h-2 replied:
> why not call it update_proxy?

it doesn't normally update another proxy, but the underlying data structure.

> This seems to be an event, in different languages I would write on_update 

I like `on_update()`! I will change it!

marehr commented on include/seqan3/alphabet/detail/alphabet_proxy.hpp:
```diff
@@ -0,0 +1,292 @@
+    }
+
+    //!\brief Assigment from the emulated type. This function triggers the specialisation in the derived_type.
+    constexpr derived_type & operator=(alphabet_type const & c) noexcept
```
Is this noexecpt? Because your all your examples, like reference_proxy_type, don't have noexecpt

h-2 replied:
They had `constexpr` which was sufficient, but I have now started adding `noexcept` everywhere, too.

marehr commented on include/seqan3/range/container/bitcompressed_vector.hpp:
```diff
@@ -125,12 +131,21 @@ class bitcompressed_vector
         ranges::semiregular_t<internal_proxy_type> internal_proxy;
 
+        //!\brief Update the sdsl-proxy.
+        constexpr void impl()
```
noexcept?

marehr commented on test/unit/alphabet/alphabet_test.cpp:
```diff
@@ -51,6 +52,24 @@
+    using base_t = alphabet_proxy<alphabet_proxy_test, dna4>;
+    friend base_t;
+
+    constexpr void impl()
```
noexcept?

marehr commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -53,10 +51,14 @@
  * \implements seqan3::nucleotide_concept
  * \implements seqan3::detail::constexpr_alphabet_concept
+ * \implements seqan3::trivially_copyable_concept
+ * \implements seqan3::standard_layout_concept
```
shouldn't this be in the `base_class`? But okay, maybe someone who uses the `alphabet_base` or `nucleotide_base` and does stupid stuff could mess this requirement up.

Maybe we should subsume `seqan3::detail::constexpr_alphabet_concept`, `seqan3::trivially_copyable_concept` and `seqan3::standard_layout_concept` as `trivial_alphabet_concept` which I think all of our alphabets satisfy.

h-2 replied:
> shouldn't this be in the base_class?

But the base_class itself doesn't guarantee this, a derived type could always violate this.

> Maybe we should subsume seqan3::detail::constexpr_alphabet_concept, seqan3::trivially_copyable_concept and seqan3::standard_layout_concept as trivial_alphabet_concept which I think all of our alphabets satisfy.

Yes, I thought about this, too. Let's discuss it separately.

marehr replied:
Cool! Put it to the TODO-List

marehr commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -248,123 +128,70 @@ struct dna15
+            {
+                ret[         rank_to_char[rnk] ] = rnk;
+                ret[to_lower(rank_to_char[rnk])] = rnk;
+            }
```
I'm surprised that this even works. I had problems in the past to use variables in the same class at declaration time.

h-2 replied:
?

marehr commented on include/seqan3/alphabet/nucleotide/nucleotide_base.hpp:
```diff
@@ -47,17 +47,26 @@ namespace seqan3
+ *
+ * The derived type needs to define only the following table as static member variable:
+ *
+ *   * `static std::array<THAT_TYPE, value_size> complement_table` that defines for every possible rank value
```
`THAT_TYPE` is not further described, also somehow the other tables `rank_to_char` and `char_to_rank` are missing.

h-2 replied:
> THAT_TYPE is not further described,

How would you change it` `derived_type`?

> also somehow the other tables rank_to_char and char_to_rank are missing.

They are provided by nucleotide_base to the derived class does not need to offer them.

marehr replied:
> > THAT_TYPE is not further described,
> 
> How would you change it `derived_type`?

Maybe like this:

```c++
// * The derived type (i.e. THAT_TYPE) needs to define only the following table as static member variable:
```

or 

```c++
 * The derived type needs to define only the following table as static member variable:
 *
 *   * `static std::array<THAT_TYPE, value_size> complement_table` that defines for every possible rank value
 *     the corresponding complement.
 * 
 * where THAT_TYPE is the name of the derived type which inherits seqan3::nucleotide_base (e.g. seqan3::dna4, ...)
```

> 
> > also somehow the other tables rank_to_char and char_to_rank are missing.
> 
> They are provided by nucleotide_base to the derived class does not need to offer them.

? dna4, dna5,... still provide rank_to_char and char_to_rank and are not provided by nucleotide_base. Maybe I understand your answer wrong, but I think we should mention that nucleotide_base inherits the requirements of alphabet_base ans must be provided by the derived_type.

marehr commented on test/unit/alphabet/nucleotide_test.cpp:
```diff
@@ -210,59 +166,59 @@ TYPED_TEST(nucleotide, explicit_conversion)
 {
     dna4_vector v;
-    v.resize(5, dna4::A);
+    v.resize(5, 'A'_dna4);
```
There is a test missing that single quote literals are really have the rank representation, i.e. `'A'_dna4` == `dna4{}.assign_rank(0)`.

marehr commented on include/seqan3/alphabet/composition/cartesian_composition.hpp:
```diff
@@ -86,118 +201,125 @@ namespace seqan3
-        using invoke = std::integral_constant<bool, implicitly_convertible_to_concept<T, type>>;
-    };
+        //!\brief Store a pointer to the parent object so we can update it.
+        cartesian_composition *parent;
```
std::reference_wrapper ?

h-2 commented on include/seqan3/alphabet/detail/alphabet_base.hpp:
```diff
@@ -0,0 +1,251 @@
+     *
+     * Guaranteed not to throw.
+     */
+    constexpr derived_type & assign_char(std::conditional_t<std::Same<char_type, void>, char, char_type> const c) noexcept
```
semi_alphabets don't have char types. So specialisations of alphabet_base  that are only semi-alphabet need to set this as void to prevent an overload of `assign_char` from being generated. Unfortunately even though the constraint would prevent the template from being instantiated when the type is void, the template would still be ill-formed, because void is not allowed for parameters. 

Therefore I need to wrap the type in the conditional.

h-2 commented on include/seqan3/alphabet/aminoacid/aminoacid_base.hpp:
```diff
@@ -0,0 +1,139 @@
+template <typename derived_type, auto size, typename char_t = char>
+class aminoacid_base : public alphabet_base<derived_type, size, char_t>
+{
+private:
```
Constructors are private to prevent wrong inheritance from the CRTP-base... we discussed this already, right?

marehr commented on include/seqan3/alphabet/composition/detail.hpp:
```diff
@@ -0,0 +1,203 @@
+{
+    //!\brief The returned type when invoked.
+    template <typename type>
+    using invoke = std::integral_constant<bool, std::is_constructible_v<type, T>>;
```
std::bool_constant

marehr commented on include/seqan3/alphabet/composition/detail.hpp:
```diff
@@ -0,0 +1,203 @@
+//!\cond
+    requires (detail::constexpr_alphabet_concept<alternative_types> && ...) &&
+             (sizeof...(alternative_types) >= 2)
+             //TODO same char_type
```
open TODO

marehr commented on include/seqan3/core/metafunction/pack.hpp:
```diff
@@ -0,0 +1,44 @@
+
+//!\brief A type trait that indicates whether the first template argument is contained in the remaining.
+template<typename target_t, typename ...pack>
+struct type_in_pack : std::false_type {};
```
this file should be merged/unified with the proposal #570 

Maybe `parameter_pack_contains` or something; It should match the naming scheme which will be decided in  #570!

h-2 commented on include/seqan3/core/tuple_utility.hpp:
```diff
@@ -199,3 +199,26 @@ constexpr auto tuple_pop_front(tuple_t && t)
+inline bool constexpr is_element_of = false;
+
+//!\brief Value metafunction that returns whether a type is an element type of a tuple. [Overload to prevent recursion]
+template <tuple_like_concept tuple_t>
```
a tuple can be an element of another tuple, the important restriction is that they can't be the same tuple.

h-2 commented on include/seqan3/core/tuple_utility.hpp:
```diff
@@ -199,3 +199,26 @@ constexpr auto tuple_pop_front(tuple_t && t)
+
+//!\brief Value metafunction that returns whether a type is an element type of a tuple. [Base template]
+template <typename ...t>
+inline bool constexpr is_element_of = false;
```
Just realised that all code parts that used this have been refactored away again so I am just removing it.

h-2 commented on include/seqan3/io/stream/char_operations.hpp:
```diff
@@ -0,0 +1,130 @@
+
+//!\brief Auxilliary table for seqan3::to_upper.
+template <typename char_type>
+inline std::array<char_type, detail::size_in_values_v<char_type>> constexpr to_upper_table
```
💁 well what do you want to call it?

marehr replied:
I think the comment is that it is not defined in this commit and the previous ones. So not sure how this is defined.

joergi-w commented on include/seqan3/alphabet/structure/wuss.hpp:
```diff
@@ -426,7 +290,7 @@ typedef wuss<51> wuss51;
 // ------------------------------------------------------------------
 
-namespace seqan3::literal
+namespace seqan3
```
Okay, as long as you keep that in mind.

h-2 commented on include/seqan3/core/concept/core_language.hpp:
```diff
@@ -143,6 +154,19 @@ template <typename t>
+template <typename t>
+concept char_concept = std::Integral<t> &&
+                       (std::Same<t, char> || std::Same<t, unsigned char> || std::Same<t, signed char> ||
+                       std::Same<t, char16_t> || std::Same<t, char32_t> || std::Same<t, wchar_t>);
```
I have changed it to:
```cpp
template <typename t>
concept char_concept = std::Integral<t> &&
                       (std::Same<t, char> || std::Same<t, unsigned char> || std::Same<t, signed char> ||
#ifdef __cpp_char8_t
                        std::Same<t, char8_t> ||
#endif
                        std::Same<t, char16_t> || std::Same<t, char32_t> || std::Same<t, wchar_t>);
```

h-2 commented on include/seqan3/core/concept/core_language.hpp:
```diff
@@ -183,4 +207,18 @@ template <typename t>
+ */
+//!\cond
+template <typename t, typename u>
+concept weakly_assignable_concept = std::is_assignable_v<t, u>;
```
All the concepts that compare two types and where added by Eric Niebler also check `std::CommonReference` for those types in addition to the actual check. But this is not what we want to check for, we are only interested in the actual thing.

h-2 commented on include/seqan3/io/alignment_file/detail.hpp:
```diff
@@ -149,8 +149,8 @@ char compare_aligned_values(reference_char_type const reference_char,
 //!\cond
-    requires std::EqualityComparableWith<gap, value_type_t<ref_seq_type>> &&
-             std::EqualityComparableWith<gap, value_type_t<query_seq_type>>
+    requires std::detail::WeaklyEqualityComparableWith<gap, reference_t<ref_seq_type>> &&
```
See the previous point. There is some controversy and open propsals regarding this situation so I left in `std::detail` for now. But, yes, if we won't get it in `std::` than we need to move it out.

h-2 commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -68,168 +70,46 @@ namespace seqan3
-    constexpr bool operator>(dna15 const & rhs) const noexcept
+    //!\brief Allow implicit construction from dna/rna of the same size.
+    template <std::Same<rna15> t>    // Accept incomplete type
+    constexpr dna15(t const & r)
```
`constexpr` implies `noexcept` up until c++17, but you are right, I should add it so it's safe for c++20.

h-2 commented on include/seqan3/io/alignment_file/detail.hpp:
```diff
@@ -242,7 +242,8 @@ std::string get_cigar_string(ref_seq_type && ref_seq,
  * \sa seqan3::aligned_sequence_concept
  */
-template<tuple_like_concept alignment_type>
+// template<tuple_like_concept alignment_type> // SOME BUG SOMEWHERE
```
This has been resolved, thanks for spotting it.

h-2 commented on include/seqan3/alphabet/detail/alphabet_proxy.hpp:
```diff
@@ -0,0 +1,292 @@
+
+#if 0 // this currently causes GCC ICE in cartesian_composition test
+    constexpr alphabet_type complement() const noexcept
+        requires nucleotide_concept<alphabet_type>
```
I tried a whole lot of stuff, I will just investigate this over the holidays!

marehr commented on include/seqan3/alphabet/gap/gap.hpp:
```diff
@@ -103,10 +105,8 @@ struct gap
     //!\brief Assign from a numeric value (no-op, since gap has only one character).
     //!\param i not used, since gap has only one character
-    constexpr gap & assign_rank([[maybe_unused]] rank_type const i) /*noexcept*/
+    constexpr gap & assign_rank([[maybe_unused]] rank_type const i) noexcept
```
This is not in this changes, but you did not remove `gap::GAP`, everywhere else you removed them and also `dna4::UNKNOWN` which I think is handy to have and something that can be represented by.

Or introduce `'-'_gap`

h-2 replied:
> Or introduce '-'_gap

There is a follow-up TODO for this.

h-2 commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -68,168 +70,46 @@ namespace seqan3
+    //!\brief Befriend seqan3::nucleotide_base.
+    friend base_t;
+    //!\cond \brief Befriend seqan3::alphabet_base.
+    friend base_t::base_t;
```
![image](https://user-images.githubusercontent.com/2996863/49151198-2b34c780-f307-11e8-8b5a-83360a54bb02.png)

h-2 commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -54,233 +51,65 @@
  * The alphabet may be brace initialized from the static letter members. Note that you cannot
  * assign the alphabet by using letters of type `char`, but you instead have to use the
- * function seqan3::dna4::assign_char().
+ * function seqan3::'a'_dna4ssign_char().
```
search anbd replace error

h-2 commented on include/seqan3/alphabet/nucleotide/dna4.hpp:
```diff
@@ -54,233 +51,65 @@
-    constexpr dna4 complement() const noexcept
+    //!\brief Allow implicit construction from dna/rna of the same size.
+    template <std::Same<rna4> t>    // Accept incomplete type
+    constexpr dna4(t const & r)
```
added

h-2 commented on include/seqan3/alphabet/detail/alphabet_base.hpp:
```diff
@@ -0,0 +1,251 @@
+
+    //!\name Comparison operators
+    //!\{
+    friend constexpr bool operator==(derived_type const & lhs, derived_type const & rhs) noexcept
```
This is a function. It means something different. See #592

h-2 commented on include/seqan3/alphabet/aminoacid/aminoacid_base.hpp:
```diff
@@ -0,0 +1,139 @@
+ *                      seqan3::semi_alphabet_concept).
+ */
+template <typename derived_type, auto size, typename char_t = char>
+class aminoacid_base : public alphabet_base<derived_type, size, char_t>
```
already changed

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+     * \{
+     */
+    //!\brief The integer representation of a quality score assignable with =operator.
+    using phred_type = int8_t;
```
yes, everyone has a phred_type, although not necessarily as member.

marehr commented on include/seqan3/alphabet/aminoacid/aminoacid_base.hpp:
```diff
@@ -0,0 +1,139 @@
+     */
+    //!\{
+    static derived_type constexpr A          = assign_char(derived_type{}, 'A');
+    static derived_type constexpr B          = assign_char(derived_type{}, 'B');
```
Yeah I meant this.

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+    constexpr quality_base & operator=(quality_base const &) = default;
+    constexpr quality_base & operator=(quality_base &&) = default;
+
+    constexpr quality_base(phred_type const p) noexcept
```
Yes, but it's a special interface that refers to neither rank nor char, it refers to the phred. Before this PR phred where actually initialised by ranks which is really broken. One could think about making a phred strong type and an int literal that returns this strong type, but that would be a feature change and not part of this PR.

marehr replied:
Please add it to future TODOs and reference this comment. Thank you!

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+ *                      seqan3::semi_alphabet_concept).
+ */
+template <typename derived_type, auto size, typename char_t = char>
+class quality_base : public alphabet_base<derived_type, size, char_t>
```
All our quality stuff is phred based, there is no difference. It's not usable by `qualified` because `qualified` is a cartesian composition. This is the same as `nucleotide_base` which is also not used by all nucleotide composition, just the "core alphabets".

marehr replied:
Thank you for clarifying! I would still propose the name change, because I changed the test name in #523 from `quality_test_template` to `phred_test_template`, because it was confusing that a test called `quality` in a subfolder named `quality` wasn't used `qualified` tests. I think you even remarked this in #535.

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+            phred_type sup = min + value_size;
+
+            // map too small to smallest possible
+            for (int64_t i = std::numeric_limits<phred_type>::lowest(); i < min ; ++i)
```
There is a smal error, because it should be `<= max`. If that is used, it will result in an overflow and endless loop if we stay in type. (the upper border cannot be part of the type itself by definition)

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+
+            // map too small to smallest possible
+            for (int64_t i = std::numeric_limits<phred_type>::lowest(); i < min ; ++i)
+                ret[static_cast<rank_type>(i)] = - derived_type::offset_phred + min;
```
> why static cast into rank_type? and not size_t?

Because underflow will result in values out of the range

> what's up with the order?

has changed already

marehr replied:
> > why static cast into rank_type? and not size_t?
> 
> Because underflow will result in values out of the range

I know, but is rank_type always unsigned? Has it the same size as phred_type? As someone who only sees this code it is not apparent which assumptions hold here. And therefore the confused question from me. It would be more clear if you use `std::make_unsigned_t<phred_type>`.

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+        {
+            std::array<rank_type, 256> ret{};
+
+            phred_type min = derived_type::offset_phred;
```
its gone

h-2 commented on include/seqan3/alphabet/quality/quality_base.hpp:
```diff
@@ -0,0 +1,195 @@
+            std::array<rank_type, 256> ret{};
+
+            phred_type min = derived_type::offset_phred;
+            phred_type sup = min + value_size;
```
its gone

h-2 commented on include/seqan3/alphabet/aminoacid/aminoacid_base.hpp:
```diff
@@ -0,0 +1,139 @@
+{
+private:
+    //!\brief Type of the base class.
+    using base_t = alphabet_base<derived_type, size, char_t>;
```
I am confused. Didn't we agree last time, that we can use this? I thought `_type` is only for exposed member types, i.e. public?

marehr replied:
True, but the confusing thing is that you call some classes `derived_type`, but the base `base_t`,

h-2 commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -68,168 +70,46 @@ namespace seqan3
  */
-
-struct dna15
+class dna15 : public nucleotide_base<dna15, 15>
```
Yes, everything that is not an aggregate POD type should be `class`.

h-2 commented on include/seqan3/alphabet/nucleotide/all.hpp:
```diff
@@ -55,104 +55,82 @@
+ * |   C      | Cytosine               |                              C       |                              C       |                              C       |                              C       |                              C       |                              C       |
+ * |   G      | Guanine                |                              G       |                              G       |                              G       |                              G       |                              G       |                              G       |
+ * |   T      | Thymine (DNA)          |                              T       |                              T       |                              T       | <div style="color:LightGrey">U</div> | <div style="color:LightGrey">U</div> | <div style="color:LightGrey">U</div> |
+ * |   U      | Uracil (RNA)           | <div style="color:LightGrey">T</div> | <div style="color:LightGrey">T</div> | <div style="color:LightGrey">T</div> |                              U       |                              U       |                              U       |
```
> use <span ...> instead of <div...>, because you most likely don't want block elements in a table cell.

I have no clue what you are talking about, I just copy'n'pasted this. If span is better I can use that.

marehr replied:
xD :+1:

h-2 commented on include/seqan3/alphabet/nucleotide/all.hpp:
```diff
@@ -55,104 +55,82 @@
+ *   5. to actually save space from using smaller alphabets, you need a compressed container (e.g.
+ *      seqan3::bitcompressed_vector)
  *
  * \par Printing and conversion to char
```
Yes, but I didn't touch this line :innocent:

marehr replied:
:rofl: Only like 8000 line changes which are totally equivalent and non-api breaking :rabbit: 

Put it to the future TODOs: doxygen: convert `\par` etc into `markdown`

h-2 commented on include/seqan3/alphabet/nucleotide/all.hpp:
```diff
@@ -55,104 +55,82 @@
- * since obviously some bases are not available. When converting to seqan3::dna5 or seqan3::rna5, non-canonical bases
+ * When assigning from `char` or converting from a larger nucleotide alphabet to a smaller one, *loss of information*
+ * can occur since obviously some bases are not available. When converting to seqan3::dna5 or seqan3::rna5,
+ * non-canonical bases
```
I tried to minimise the line-diff

h-2 commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -68,168 +70,46 @@ namespace seqan3
-
-    //!\copydoc seqan3::dna4::value_size
-    static constexpr rank_type value_size{15};
+private:
```
Hm, I thought it was always part of our style to still write this? I don't remember, but I have always put them there to be clear about it.

h-2 commented on include/seqan3/alphabet/nucleotide/dna15.hpp:
```diff
@@ -248,123 +128,70 @@ struct dna15
+            {
+                ret[         rank_to_char[rnk] ] = rnk;
+                ret[to_lower(rank_to_char[rnk])] = rnk;
+            }
```
The problem is that they are initialised with different values before the loop is called. In the end, I found that it doesn't help to factor this out in the current design.

marehr replied:
Okay, so you thought about it :+1:

h-2 commented on include/seqan3/alphabet/gap/gap.hpp:
```diff
@@ -117,32 +117,32 @@ struct gap
     }
 
-    constexpr bool operator!=(gap const &) const noexcept
+    friend constexpr bool operator!=(gap const &, gap const &) noexcept
```
We want to implement all comparison operators as friends, but we cannot constrain them because of bugs.

marehr replied:
Ah okay, thank you!

h-2 commented:
> I think there should have been a discussion

>    whether we allow 'A'_dna4 literals (in respect to usability, does a user easily understand the difference between 'A'_dna4 and "A"_dna4?),

Users understand the difference between "A" and 'A', too, right? Or to put it differently, not all users understand the differences, but they are no less confused by the enum notation. This is closer to just assigning 'A' or just assigning "ACGT" than anything else.

>    that we break API and remove dna4::A and

We have no API stability, yet, so nothing is being broken :wink: 

I have discussed this design change with @rrahn and I humbly assumed that since I basically designed this module and had to refactor it now and had to do broken stuff like `this_deferred_type` any other solution would be better.

>    that we reorder the rank representation (now lexicographical, before logical? order).

I discussed this design change with @smehringer. I can explain the reasoning again in person, but I would really not want to delay merging this because of that :pray: 

> The CRTP refactoring looks good to be honest and I think was the right step in a better direction.

Thanks for the thorough review!

marehr commented:
@h-2 In reply to https://github.com/seqan/seqan3/pull/535#issuecomment-442456733:  None of my remarks are any reason to not approve your PR, just remarks for better communication in the future. (This would be stuff to discuss in the core team in the future, but since this was not established when you wrote this, I'm the last person to block the changes.)

The reason I did not approve yet, is that I'm not done with the review :love_letter:

marehr commented on test/unit/range/view/view_translation_test.cpp:
```diff
@@ -46,15 +46,13 @@
 using namespace seqan3;
-using namespace seqan3::literal;
-using namespace seqan3::view;
+// using namespace seqan3::view;
```
commented out

marehr replied:
roger

marehr commented:
unit test: 

```c++
 31/117 Test  #31: unit/alphabet/gap/gapped_test .......................................***Failed    0.00 sec
Running main() from /home/travis/build/seqan/seqan3-build/vendor/googletest/googletest/src/gtest_main.cc
[==========] Running 3 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 3 tests from gapped_test
[ RUN      ] gapped_test.initialise_from_component_alphabet
[       OK ] gapped_test.initialise_from_component_alphabet (0 ms)
[ RUN      ] gapped_test.assign_from_component_alphabet
[       OK ] gapped_test.assign_from_component_alphabet (0 ms)
[ RUN      ] gapped_test.fulfills_concepts
/home/travis/build/seqan/seqan3/test/unit/alphabet/gap/gapped_test.cpp:101: Failure
Value of: (std::is_pod_v<alphabet_t>)
  Actual: false
Expected: true
/home/travis/build/seqan/seqan3/test/unit/alphabet/gap/gapped_test.cpp:102: Failure
Value of: (std::is_trivial_v<alphabet_t>)
  Actual: false
Expected: true
[  FAILED  ] gapped_test.fulfills_concepts (0 ms)
[----------] 3 tests from gapped_test (0 ms total)
[----------] Global test environment tear-down
[==========] 3 tests from 1 test case ran. (0 ms total)
[  PASSED  ] 2 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] gapped_test.fulfills_concepts
 1 FAILED TEST
```

documentation:

```c++
2/2 Test #2: doc_dev_test .....................***Failed   41.92 sec
/home/travis/build/seqan/seqan3/include/seqan3/range/container/bitcompressed_vector.hpp:120: warning: explicit link request to 'impl()' could not be resolved /home/travis/build/seqan/seqan3/include/seqan3/range/container/bitcompressed_vector.hpp:120: warning: explicit link request to 'impl()' could not be resolved
```

h-2 commented:
superseded by #595
